### DISCUSSION TITLE
Use of the USB-Serial port by the application

### DISCUSSION THREAD

**Discussion by costaelder** [micropython/micropython#9936] [Using MicroPython] [answered]
My target processor board is the Raspberry Pi Pico.

I have searched in the Internet and have seen some posts in the Micropython forum but I was unable to find a definitive answer to my question, so here it goes.

I would like to use the USB serial port of the Pico to communicate with a PC during the execution of my Micropython application. I have seen a few posts on the subject but it still unclear to me if it is possible doing it with the Pico. I am designing an application that in the future will need this communication (initially it will work standalone, with reduced funcionality). I may reserve a pair of pins for a UART to connect to the PC via a separate USB-Serial converter, but I am quickly running out of pins so I would like to do some quick communication tests to check if I may save these two pins for other functions.

On a somewhat off-topic subject, is there some protocol I could use for this communication? I want to avoid reinventing the wheel coding one myself. I remember seeing some simple protocols to use via asynchronous channels, but I have not been able to find any, probably because of bad choice of keywords in the search.

**Comment by jimmo** [comment, marked as answer]
Yes, anything you read/write from sys.stdin/sys.stdout in MicroPython will be transmitted by default over the virtual USB UART. You can write a program on your PC that reads/writes from the serial port. e.g. in Python using pyserial.

https://github.com/BrianPugh/belay might be worth checking out too.

**Reply by costaelder** [reply]
Hello, jimmo.

Thank you for your answer.

Re belay, I will be looking at it deeper. I am not sure if it would work for this particular project, but I think it might work for some other projects I have in my queue, where I was considering [circuit python](https://learn.adafruit.com/circuitpython-libraries-on-any-computer-with-raspberry-pi-pico).

Any suggestion on a simple protocol for the serial port? I have looked for an implementation in python, I thought of xcp, but it looks like overkill for my application.

**Reply by jimmo** [reply]
I think it depends on your project, and in particular whether it's command/response or both sides can initiate.

I have seen people implement HTTP-over-serial with MicroPython :)

The approach taken by belay and mpremote is Python-commands-over-serial. i.e. if you want the board to execute the "foo()" function, just tell it to run foo() and wait for the result, which will be formatted as a Ptyhon object that can be loaded with https://docs.python.org/3/library/ast.html#ast.literal_eval.

**Comment by emcek** [comment]
I got similar problem, but I think `belay` isn't answer in my case.

I looking for (at best) non-blocking both way communication Pico <=> PC via serial. I need send something to Pico (i.e. print at LCD) and allow Pico to send to PC events from LCD touch screen or onboard buttons.

'belay' looks like communication is initiate by PC only: printing something at LCD or return value from some function executed at Pico.

**Reply by jimmo** [reply]
You can adapt pyboard.py for this purpose.

**Reply by emcek** [reply]
thx I will take a look!

**Reply by emcek** [reply]
looks like `ProcessPtyToTerminal` or `ProcessToSerial` can help solve my case, right?

**Reply by jimmo** [reply]
I was imagining that you could use the higher-level Pyboard class directly.
